### PR TITLE
Preservation Blog Post

### PR DESCRIPTION
**Pull request by tpendragon**
Largely written as a response to https://github.com/pulibrary/figgy/issues/2884

@escowles @jpstroop @hackmastera @eliotjordan @jrgriffiniii @christinach @tampakis Can y'all look at this? I put in a lot of "we" here, so I want to make sure I'm not misrepresenting Princeton in any way.

@rosy1280 This is in response to your ticket, and you've had great notes already, so could you also take a peek?

### CODE REVIEW

File: _posts/2019-04-16-preservation.md [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+This means a requirement is that I'm able to copy a single directory and get
+everything I need to re-ingest that object and display it. External references
+may not come in, but the hierarchy will.
+
+So a Volume will be nested underneath its Multi-Volume Work, and a Page will be
+nested underneath its Book. This is important, because if one decides against
+this then the pros/cons of the following cases may change significantly.
+
+### Assumptions
```
Comment (escowles) [inline _posts/2019-04-16-preservation.md:35]: I think we should add something to this section about being able to verify that binary files are intact using their stored checksums.

File: _posts/2019-04-16-preservation.md [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+   corrupted by a race condition:
+   1. Preservation starts. Pages start to persist into a bag.
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved in its spot in the bag, updating manifest.
+   5. Initial preservation ends. Updates manifest, but has out-of-date checksum
+      information.
+1. Files such as `bagit.txt` and `bag-info.txt` do nothing for the preservation
+   itself. They're simply there to satisfy the specification, and will result in
+   using up (admittedly a small amount) of space for no reason.
```
Comment (escowles) [inline _posts/2019-04-16-preservation.md]: I'm not sure I agree with this point — they aren't important to us, but they identify the package as a bag, and include metadata in a format that other people are able to easily work with.  So they are part of letting people (including us in the future) work with the package.

Comment (tpendragon) [reply _posts/2019-04-16-preservation.md]: Yeah, I'll just remove it. I'm not a huge fan of files for the sake of files, but it's not a strong argument.

File: _posts/2019-04-16-preservation.md (hunk 1 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:35]: Added a small sentence - look good?

**Comment by jpstroop** [comment]
👍 This raises interesting questions about OCFL for me. Simply put: I've recently (probably late to the party) started to realize that if you're just treating the cloud as a big farm of VMs, then you're doing it wrong, at least in terms of economy of scale, elasticity, etc, and you really need to drink the kool-aid (and think at a higher level of abstraction, e.g., about buckets, not filesystems or directories, etc.). And given the trend toward the cloud, how relevant is a POSIX-based solution?

Or, is it more complicated than that? Should we be using cloud-based storage for our production systems, and periodically dumping periodic snapshots (versions?) back to a less opaque/abstract, more traditional storage solution? Pessimistic locking seems like less of an inconvenience if we were just making, say, quarterly incremental dumps to a regular filesystem somewhere.

**Comment by tpendragon** [comment]
> Or, is it more complicated than that? Should we be using cloud-based storage for our production systems, and periodically dumping periodic snapshots (versions?) back to a less opaque/abstract, more traditional storage solution? Pessimistic locking seems like less of an inconvenience if we were just making, say, quarterly incremental dumps to a regular filesystem somewhere.

Yes. Or don't even snapshot things to traditional storage - automated means exist to migrate data between cloud providers at periodic intervals without us touching anything and using their pipes. We don't even need a special preservation system if we were on GCS in the first place - marking it "cloud" could just move the binary content over to the preservation bucket (for free), which has versioning turned on and a lifecycle rule to auto-coldline versions.

We're not there yet, though.

File: _posts/2019-04-16-preservation.md [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+   1. Preservation starts. Pages start to persist into a bag.
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved in its spot in the bag, updating manifest.
+   5. Initial preservation ends. Updates manifest, but has out-of-date checksum
+      information.
+
+#### OCFL
+
+Implementation notes for OCFL recommend things such as pair-trees for storing
```
Comment (rosy1280) [inline _posts/2019-04-16-preservation.md]: It says you can do this not that you need to do this. We know it's a use case many currently have so we addressed it.

We'll clear this up that it's one potential way of storing in the implementation notes.

File: _posts/2019-04-16-preservation.md [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+   browsers, and will be understood by partner institutions.
+1. This structure just needs a file system to store itself and support
+   versioning, there's no need for an object store's feature-set.
+1. We can de-duplicate as part of generating the `inventory.json` if we'd like
+   to, using up less space than the other options.
+
+**Cons**:
+
+1. The inventory file acts as a pessimistic lock in the same way as BagIt's
+   manifest file.
```
Comment (rosy1280) [inline _posts/2019-04-16-preservation.md:188]: Is this a problem you face with OCFL and BagIt or with any kind of versioning?

File: _posts/2019-04-16-preservation.md [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+1. We have to prevent future changes to a child from persisting to preservation
+   storage until this large change is done. Otherwise, the inventory will be
+   corrupted by a race condition, and versions will be unable to be properly
+   calculated:
```
Comment (rosy1280) [inline _posts/2019-04-16-preservation.md:192]: A problem with many versioning systems is determining what constitutes a version. For example if my title lacks a period at the end of it and I add the period is that a new version? If I have a massive data set a researcher is working on do I really want to keep a new copy each time it changes? Or do I want to version it at moments that make sense in the researcher's lifecycle (after a publication, before major grant work, etc.). OCFL provides freedom for the institution to make those decisions. 

Simply put, OCFL does not prescribe when a version is created, we leave that up to the software that sits on top of the storage root.

File: _posts/2019-04-16-preservation.md [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+      before we start writing things if files exist to reduce the time we need
+      to lock.
+
+#### Google Cloud Storage
+
+A small note about GCS - I'm going to call paths "directories", but in GCS it's
+  all just naming semantics. `foo/bar/txt.yml` isn't a file in the directory
+  `bar`, it's a file that has a name which happens to have slashes in it. The
+  GCS tools just visualize them as directories to be nice.
+
```
Comment (rosy1280) [inline _posts/2019-04-16-preservation.md:210]: This is also how OCFL does things. Files are their hashes with a name so it's easier for humans.

File: _posts/2019-04-16-preservation.md [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+**Cons**:
+
+1. Prioritizing parallelization makes de-duplicating infeasible.
+2. The format's not standard. It's easy to look at, but we can't point at a
+   specification. At least it should be easy to convert to any of the other
+   formats if we'd like to.
+3. Migrating to a different platform is potentially more than a single copy
+   operation - but at least GCS validates file integrity as part of the
+   platform.
```
Comment (rosy1280) [inline _posts/2019-04-16-preservation.md:237]: Are you also locked into GCS? Meaning what would migration to a new storage infrastructure look like in the future? How would you capture all of the changes you've made when you migrate to this new system since you're relying on GCS for versioning?

File: _posts/2019-04-16-preservation.md [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+
+1. Create a new `<book-id>/v2 directory.
+2. Create a `v2/inventory.json` file with with a new `version` entry that has
+   nothing in the `state` property.
+3. Create a new `v2/inventory.json.sha512`
+4. Copy the new `inventory.json` and `inventory.json.sha512` to the top
+   directory.
+
+There are unlikely to be locking problems, because nothing happens to the object
+post-delete.
```
Comment (rosy1280) [inline _posts/2019-04-16-preservation.md:261]: This is one way to delete an object, if you're ok with also keeping it. If you need to completely delete it for legal reasons there is another way outlined in the implementation notes. 

I wonder if PUL's collections would need this though (Emory's collections are more modern and this is a use case we've already had to deal with even lacking a comprehensive access system).

**Review by rosy1280 — COMMENTED** [review]
Overall I noticed diversity of storage and multiple copies were not addressed in this blog post. Is PUL considering these issues?

**Comment by tpendragon** [comment]
> Overall I noticed diversity of storage and multiple copies were not addressed in this blog post. Is PUL considering these issues?

Last I checked the intention was three copies: our local storage+backup solutions (Isilon), GCS Preservation, and a copy in AWS. The mechanics of copying over to AWS haven't been decided yet. I don't say this with authority though.

File: _posts/2019-04-16-preservation.md (hunk 8 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:261]: We've had to do it in the past, is my understanding, it just wasn't the use case here.

**Comment by rosy1280** [comment]
Presumably you'll use the internal audit trails provided by each of those systems?

File: _posts/2019-04-16-preservation.md (hunk 7 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:237]: I'm of the opinion we're not. Their versioning API is clearly documented and easily accessible from a variety of tools maintained both by google and third parties. If we wanted move to OCFL we could easily create a manifest, if we moved to something else that supported versions we could put them in there.

File: _posts/2019-04-16-preservation.md (hunk 6 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:210]: Yeah, but OCFL isn't a platform so I can't say a similar thing right? Like, if I delete foo/bar/txt.yml, then foo/bar is gone. The spec says I should just ignore foo/bar for OCFL, but can I say it won't be there?

File: _posts/2019-04-16-preservation.md (hunk 5 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:192]: I'm not sure how I can put that here. In our case we want every change post-complete to not be lost - every versioning system for a preservation store you're not using as your database has that feature right? It's just whenever you decide to preserve its state again. So I guess the race condition is more "prevent future versions from being created until this one is done"? Does that sound better?

File: _posts/2019-04-16-preservation.md (hunk 4 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:188]: OCFL and BagIt. The problem is there's a single file which needs to be updated to represent a logical unit made up of several resources. Every write to any resource results in having to write that same file. If we rely on GCS' versioning, or any system which doesn't have a single file inventory, then there's no point of contention and I can append pieces of the preservation store with no worry.

File: _posts/2019-04-16-preservation.md (hunk 3 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md]: :+1: Apologies, I'll change the wording to "Implementation notes for OCFL provides options for things such as...", is that a better representation?

**Comment by tpendragon** [comment]
> Presumably you'll use the internal audit trails provided by each of those systems?

I assume you mean similar to how the OCFL spec has an optional "logs" directory for such things? If so, I'm unsure! We haven't implemented an event log for resources yet in the vein of PREMIS events - there's a good chance they'd be preserved as resources alongside the metadata for each resource.

File: _posts/2019-04-16-preservation.md (hunk 5 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:192]: Oh! Unless you're proposing it's kosher to just mutate the files in the existing files and fix up the checksums? I guess that'd be fine. Do you think it'd be helpful to put that here? It might make this case more complex, and I'm not sure we'd do it, especially because you'd have to determine if the past files should also be mutated, since it's all de-duped, and if so how far back..

File: _posts/2019-04-16-preservation.md [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+
+**Cons**:
+
+1. The inventory file acts as a pessimistic lock in the same way as BagIt's
+   manifest file.
+1. We have to prevent future changes to a child from persisting to preservation
+   storage until this large change is done. Otherwise, the inventory will be
+   corrupted by a race condition, and versions will be unable to be properly
+   calculated:
+   1. Preservation starts. Pages start to persist into a bag.
```
Comment (hackartisan) [inline _posts/2019-04-16-preservation.md]: change `bag` to something else; `directory`?

File: _posts/2019-04-16-preservation.md [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+layout: default
+---
+
+## Preservation
+**by Trey Pendragon**
+
+Lately we've been working on preservation within
+[Figgy](https://github.com/pulibrary/figgy) and have come up with a strategy we
+think works for us. However, in doing so, we've evaluated a few options and come
+up with some assumptions which could be useful to others.
```
Comment (tpendragon) [inline _posts/2019-04-16-preservation.md:13]: Add something here about how this isn't our entire preservation story, just comparing some technical storage details.

**Comment by tpendragon** [comment]
@hackmastera @rosy1280 I added a few changes based on your comments, could you take another look?

**Comment by jpstroop** [comment]
I wonder if @zimeon might like to give this a look?

File: _posts/2019-04-16-preservation.md (hunk 6 above)
Comment (rosy1280) [reply _posts/2019-04-16-preservation.md:210]: if you want to keep foo/bar you should create foo/bar/.keep to future proof storage migrations (because another system might not keep foo/bar)

File: _posts/2019-04-16-preservation.md (hunk 7 above)
Comment (rosy1280) [reply _posts/2019-04-16-preservation.md:237]: The proposal above seems to work today. The nuances of relying on any third party storage provider for storing your audit trails is something that might need further consideration.

**Comment by rosy1280** [comment]
> I assume you mean similar to how the OCFL spec has an optional "logs" directory for such things?

GCS is tracking changes to your content via its versioning methods which presumably would be used to audit the content (i.e. is this the content i'm looking for?).  I'm asking if you will also be using versioning and audit built into your local storage environment and AWS S3.

**Comment by tpendragon** [comment]
> GCS is tracking changes to your content via its versioning methods which presumably would be used to audit the content (i.e. is this the content i'm looking for?). I'm asking if you will also be using versioning and audit built into your local storage environment and AWS S3.

Ah, yes. Probably at least, we haven't sussed out those details yet.

File: _posts/2019-04-16-preservation.md (hunk 6 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:210]: I'll add something to this section.

Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:210]: Pushed up a small paragraph. Look okay @rosy1280?

File: _posts/2019-04-16-preservation.md (hunk 7 above)
Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:237]: I added a con below this note regarding the fact that we'd have to use their version API. I think whether or not that's a good idea can probably be less a subject of the post and more some discussion on https://github.com/pulibrary/figgy/issues/2884 - does that seem okay?

**Comment by tpendragon** [comment]
@escowles @jpstroop I'd like to get this up before I forget about it. Is everything okay, or would we like to wait for more input?

File: _posts/2019-04-16-preservation.md (hunk 5 above)
Comment (escowles) [reply _posts/2019-04-16-preservation.md:192]: I think the implication is that we'd need to buffer changes and not mutate the OCFL package the way we are mutating our packages now.  I think that's a thing we should look into, but not something we need to resolve right now.

File: _posts/2019-04-16-preservation.md (hunk 8 above)
Comment (escowles) [reply _posts/2019-04-16-preservation.md:261]: Agree —  we have had to takedown and permanently delete things, but it's the rare exception, not the typical delete.

File: _posts/2019-04-16-preservation.md (hunk 4 above)
Comment (zimeon) [reply _posts/2019-04-16-preservation.md:188]: I agree that OCFL and BagIt have essentially the same property with manifest/inventory in this regard. In both cases updates to an object should be made as an atomic operation -- locked from other operations on the same object -- but it is an implementation choice whether the manifest/inventory is kept up-to-date in the middle of this overall operation. Support for a resource-within-object centric interface would suggest updating the manifest/inventory for each change which I agree would be a bottleneck. This seems a tension between preservation vs active-data approaches

File: _posts/2019-04-16-preservation.md (hunk 5 above)
Comment (zimeon) [reply _posts/2019-04-16-preservation.md:192]: I agree @escowles -- I think this is the key difference between BagIt/OCFL and your GCS-based approached. In BagIt/OCFL the object is very well-defined and bounded, there is no notion of a hierarchy of objects where something can be added inside without affected the data (manifest) for the whole. IMO this provides clarity for preservation but comes with a cost. Links from one object to other objects could provide a way around this but is hasn't been thought out for OCFL (maybe we should write up a use case for that)

Comment (tpendragon) [reply _posts/2019-04-16-preservation.md:192]: :+1: I look forward to seeing what OCFL clients do to deal with all this. It seems like the reasonable strategy is something less like "store this resource in a package" and more like "preserve this whole hierarchy please," with some pretty amazing logic for handling diffing with potentially very little access to the material upstream (in cases of expensive cloud storage - or even delayed storage with Glacier).

File: _posts/2019-04-16-preservation.md [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+This isn't our entire preservation story (there are other issues such as
+replication, management, audit trails, etc), but it compares a few options for
+our back-end storage.
+
+
+### Object Boundaries
+
+I talk about an "object" a lot below. After much discussion, we decided at
+  Princeton that we want a "directory" to consist of a top-level object (like a
+  Book) and all its children.
```
Comment (bcail) [inline _posts/2019-04-16-preservation.md:25]: Outsider comment - feel free to ignore. It'd be interesting to see some of that discussion: the pros and cons of requiring children to be nested under the parent directory.

Comment (escowles) [reply _posts/2019-04-16-preservation.md:25]: @bcail the tl;dr version is:

pro:
1. Packaging an object and its children together makes intuitive sense.
1. We expect restoration to happen at the object level, so packaging them together will make it easier to do that.
1. In disaster scenarios where Figgy isn't available, it would be much easier to work with a filesystem hierarchy that at least contained all the files and metadata for an object.  In cases where we have gone into Figgy's storage, it's very cumbersome to have to navigate around based on ids.

con:
1. What exactly is the boundary between an object and other objects?  Membership relationships usually indicate that the related things should be included, but not always (and we have modeled collection-like relationships both as members and in other ways)
1. Where do we stop? Do we want to also ingest controlled vocabulary terms? The whole controlled vocabulary hierarchy? Collections? How much of the repository do we want to duplicate, and would that serve any purpose?
1. Combining things and nesting them will make sync more challenging — if we had a 1-to-1 mapping between Figgy objects and Preservation objects, it would be simpler. But then what is the point of packaging at all? Why not just mirror the local file and metadata storage in the cloud and call it a day?
